reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/SwitchMap.kt

```diff
@@ -0,0 +1,78 @@
+import com.badoo.reaktive.base.ValueCallback
+import com.badoo.reaktive.base.subscribeSafe
+import com.badoo.reaktive.base.tryCatch
+import com.badoo.reaktive.disposable.CompositeDisposable
+import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.disposable.DisposableWrapper
+import com.badoo.reaktive.utils.atomic.AtomicBoolean
+import com.badoo.reaktive.utils.atomic.AtomicInt
+
+fun <T, R> Observable<T>.switchMap(mapper: (T) -> Observable<R>): Observable<R> =
+    observable { emitter ->
+        val disposables = CompositeDisposable()
+        emitter.setDisposable(disposables)
+        val serializedEmitter = emitter.serialize()
+
+        subscribeSafe(object : ObservableObserver<T>, ErrorCallback by serializedEmitter {
+            private val activeSourceCount = AtomicInt(initialValue = 1)
+            private val activeObserver = DisposableWrapper()
+
+            override fun onSubscribe(disposable: Disposable) {
+                disposables += disposable
+                disposables += activeObserver
+            }
+
+            override fun onNext(value: T) {
+                activeSourceCount.addAndGet(delta = 1)
+
+                serializedEmitter.tryCatch({ mapper(value) }) { innerObservable ->
+                    val innerObserver = object : ObservableObserver<R>, Disposable,
+                        ValueCallback<R> by serializedEmitter,
+                        ErrorCallback by this {
+
+                        private val disposableWrapper = DisposableWrapper()
+                        private val isCompleted = AtomicBoolean(initialValue = false)
+
+                        override fun onSubscribe(disposable: Disposable) {
+                            disposableWrapper.set(disposable)
+                        }
+
+                        override fun onComplete() {
+                            if (isCompleted.compareAndSet(expectedValue = false, newValue = true)) {
+                                if (activeSourceCount.addAndGet(delta = -1) <= 0) {
+                                    serializedEmitter.onComplete()
+                                }
+                            }
+                        }
+
+                        override val isDisposed: Boolean get() = disposableWrapper.isDisposed
+
+                        override fun dispose() {
+                            disposableWrapper.dispose()
+                            if (!isCompleted.value) {
+                                activeSourceCount.addAndGet(delta = -1)
+                            }
+                        }
+
+                    }
+                    activeObserver.set(innerObserver)
+                    innerObservable.subscribeSafe(innerObserver)
+                }
+            }
+
+            override fun onComplete() {
+                if (activeSourceCount.addAndGet(delta = -1) <= 0) {
+                    serializedEmitter.onComplete()
+                }
+            }
+
+        })
+    }
```

Comment:
I have played with the implementation and came up with the following code:
```
fun <T, R> Observable<T>.switchMap(mapper: (T) -> Observable<R>): Observable<R> =
    observable { emitter ->
        val disposables = CompositeDisposable()
        emitter.setDisposable(disposables)
        val innerDisposableWrapper = DisposableWrapper()
        disposables += innerDisposableWrapper
        val state = AtomicReference(SwitchMapState(), true)
        val serializedEmitter = emitter.serialize()

        subscribeSafe(
            object : ObservableObserver<T>, ErrorCallback by serializedEmitter {
                override fun onSubscribe(disposable: Disposable) {
                    disposables += disposable
                }

                override fun onNext(value: T) {
                    serializedEmitter.tryCatch(block = { mapper(value) }, onSuccess = ::onInnerObservable)
                }

                private fun onInnerObservable(observable: Observable<R>) {
                    val localDisposableWrapper = DisposableWrapper()

                    /*
                     * Dispose any existing inner Observable.
                     * If a previous Observable did not provide its disposable yet it will be disposed automatically later since
                     * its localDisposableWrapper is disposed.
                     */
                    innerDisposableWrapper.set(localDisposableWrapper)

                    val innerObserver =
                        object : ObservableObserver<R>, ValueCallback<R> by serializedEmitter, ErrorCallback by serializedEmitter {
                            override fun onSubscribe(disposable: Disposable) {
                                localDisposableWrapper.set(disposable)
                            }

                            override fun onComplete() {
                                checkStateFinished(state.updateAndGet { if (it.inner == this) it.copy(inner = null) else it })
                            }
                        }

                    state.update { it.copy(inner = innerObserver) }
                    observable.subscribeSafe(innerObserver)
                }

                override fun onComplete() {
                    checkStateFinished(state.updateAndGet { it.copy(isUpstreamActive = false) })
                }

                private fun checkStateFinished(state: SwitchMapState) {
                    if (state.isFinished) {
                        serializedEmitter.onComplete()
                    }
                }
            }
        )
    }

private data class SwitchMapState(
    val isUpstreamActive: Boolean = true,
    val inner: Any? = null
) {
    val isFinished: Boolean get() = !isUpstreamActive && (inner == null)
}
```

Reply:
Brilliant!